Null database is attempted to be used

When testing the addition of database testing into the CICD we got an error message that a null database was attempted to be used. This should never be the case as we should produce an error during instantiation/connection of a database and return that error instead of accepting a null database as a response. 

Full logs:
https://github.com/epag/wres-copy/actions/runs/9863767483/job/27237251392

Workflow that triggered this:
https://github.com/epag/wres-copy/blob/master/.github/workflows/postgresql.yml

TODO:
- [ ] Check for a non-null database when one is required (dont assume one is created)
- [ ] New ConnectionSupplier should throw and error when unable to connect to a database

Alright, I finished this and we return a much better error message now:

```
wres-worker-1        | 2024-07-11T13:06:14.413+0000 INFO JobStandardStreamMessenger 2024-07-11T13:06:14.411+0000 ERROR Functions Unable to create database with:
wres-worker-1        | 2024-07-11T13:06:14.414+0000 INFO JobStandardStreamMessenger java.lang.IllegalStateException: com.zaxxer.hikari.pool.HikariPool$PoolInitializa
tionException: Failed to initialize pool: FATAL: password authentication failed for user "wres_user"
wres-worker-1        | 2024-07-11T13:06:14.416+0000 INFO JobStandardStreamMessenger     at wres.io.database.ConnectionSupplier.makeConnectionPool(ConnectionSupplier.
java:91)
wres-worker-1        | 2024-07-11T13:06:14.417+0000 INFO JobStandardStreamMessenger     at wres.io.database.ConnectionSupplier.<init>(ConnectionSupplier.java:47)
wres-worker-1        | 2024-07-11T13:06:14.418+0000 INFO JobStandardStreamMessenger     at wres.Functions.openResources(Functions.java:865)
wres-worker-1        | 2024-07-11T13:06:14.420+0000 INFO JobStandardStreamMessenger     at wres.Functions.evaluate(Functions.java:114)
wres-worker-1        | 2024-07-11T13:06:14.421+0000 INFO JobStandardStreamMessenger     at wres.server.EvaluationService.lambda$startEvaluation$2(EvaluationService.j
ava:698)
wres-worker-1        | 2024-07-11T13:06:14.422+0000 INFO JobStandardStreamMessenger     at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
wres-worker-1        | 2024-07-11T13:06:14.423+0000 INFO JobStandardStreamMessenger     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExec
utor.java:1136)
wres-worker-1        | 2024-07-11T13:06:14.424+0000 INFO JobStandardStreamMessenger     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExe
cutor.java:635)
wres-worker-1        | 2024-07-11T13:06:14.425+0000 INFO JobStandardStreamMessenger     at java.base/java.lang.Thread.run(Thread.java:840)
wres-worker-1        | 2024-07-11T13:06:14.427+0000 INFO JobStandardStreamMessenger Caused by: com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed
to initialize pool: FATAL: password authentication failed for user "wres_user"
wres-worker-1        | 2024-07-11T13:06:14.428+0000 INFO JobStandardStreamMessenger     at com.zaxxer.hikari.pool.HikariPool.throwPoolInitializationException(HikariP
ool.java:584)
wres-worker-1        | 2024-07-11T13:06:14.429+0000 INFO JobStandardStreamMessenger     at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:571)
wres-worker-1        | 2024-07-11T13:06:14.430+0000 INFO JobStandardStreamMessenger     at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:98)
wres-worker-1        | 2024-07-11T13:06:14.431+0000 INFO JobStandardStreamMessenger     at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:80)
wres-worker-1        | 2024-07-11T13:06:14.432+0000 INFO JobStandardStreamMessenger     at wres.io.database.ConnectionSupplier.createDataSource(ConnectionSupplier.ja
va:125)
wres-worker-1        | 2024-07-11T13:06:14.433+0000 INFO JobStandardStreamMessenger     at wres.io.database.ConnectionSupplier.makeConnectionPool(ConnectionSupplier.
java:87)
wres-worker-1        | 2024-07-11T13:06:14.435+0000 INFO JobStandardStreamMessenger     ... 8 common frames omitted
wres-worker-1        | 2024-07-11T13:06:14.436+0000 INFO JobStandardStreamMessenger Caused by: org.postgresql.util.PSQLException: FATAL: password authentication fail
ed for user "wres_user"
wres-worker-1        | 2024-07-11T13:06:14.437+0000 INFO JobStandardStreamMessenger     at org.postgresql.Driver$ConnectThread.getResult(Driver.java:397)
wres-worker-1        | 2024-07-11T13:06:14.438+0000 INFO JobStandardStreamMessenger     at org.postgresql.Driver.connect(Driver.java:305)
wres-worker-1        | 2024-07-11T13:06:14.439+0000 INFO JobStandardStreamMessenger     at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:681)
wres-worker-1        | 2024-07-11T13:06:14.440+0000 INFO JobStandardStreamMessenger     at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:229)
wres-worker-1        | 2024-07-11T13:06:14.442+0000 INFO JobStandardStreamMessenger     at org.postgresql.ds.common.BaseDataSource.getConnection(BaseDataSource.java:
104)
wres-worker-1        | 2024-07-11T13:06:14.443+0000 INFO JobStandardStreamMessenger     at org.postgresql.ds.common.BaseDataSource.getConnection(BaseDataSource.java:
88)
wres-worker-1        | 2024-07-11T13:06:14.444+0000 INFO JobStandardStreamMessenger     at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:360)
wres-worker-1        | 2024-07-11T13:06:14.445+0000 INFO JobStandardStreamMessenger     at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:202)
wres-worker-1        | 2024-07-11T13:06:14.446+0000 INFO JobStandardStreamMessenger     at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:461)
wres-worker-1        | 2024-07-11T13:06:14.447+0000 INFO JobStandardStreamMessenger     at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:550)
wres-worker-1        | 2024-07-11T13:06:14.448+0000 INFO JobStandardStreamMessenger     ... 12 common frames omitted
```

--- a/src/wres/Functions.java
+++ b/src/wres/Functions.java
@@ -856,7 +856,16 @@ public final class Functions
              && ( Objects.isNull( Functions.database )
                   || Functions.database.isShutdown() ) )
         {
-            Functions.database = new Database( new ConnectionSupplier( sharedResources.systemSettings ) );
+            // Try to create a database connection and catch exceptions to log the reason why our connection failed
+            try
+            {
+                Functions.database = new Database( new ConnectionSupplier( sharedResources.systemSettings ) );
+            }
+            catch ( IllegalStateException ise )
+            {
+                LOGGER.error( "Unable to create database with: ", ise );
+            }
+
             // Migrate the database, as needed
             if ( Functions.database.getAttemptToMigrate() )
             {
@@ -866,7 +875,7 @@ public final class Functions
                 }
                 catch ( SQLException e )
                 {
-                    throw new IllegalStateException( "Failed to migrate the WRES database.", e );
+                    throw new InternalWresException( "Failed to migrate the WRES database.", e );
                 }
             }
         }
@@ -932,6 +941,10 @@ public final class Functions
         if ( sharedResources.systemSettings()
                             .isUseDatabase() )
         {
+            if ( Objects.isNull( Functions.database ) )
+            {
+                throw new InternalWresException( "Unable to log failure in database because connection does not exist" );
+            }
             Instant endedExecution = Instant.now();
 
             // Log both the operation and the args

--- a/wres-io/src/wres/io/database/ConnectionSupplier.java
+++ b/wres-io/src/wres/io/database/ConnectionSupplier.java
@@ -6,6 +6,7 @@ import javax.sql.DataSource;
 import com.github.marschall.jfr.jdbc.JfrDataSource;
 import com.zaxxer.hikari.HikariConfig;
 import com.zaxxer.hikari.HikariDataSource;
+import com.zaxxer.hikari.pool.HikariPool;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -80,7 +81,15 @@ public class ConnectionSupplier
         int maxPoolSize = databaseConfiguration.getMaxPoolSize();
         LOGGER.info( "Creating a database connection pool with {} connections...", maxPoolSize );
         long connectionTimeoutMs = databaseConfiguration.getConnectionTimeoutMs();
-        DataSource inner = createDataSource( maxPoolSize, connectionTimeoutMs );
+        DataSource inner;
+        try
+        {
+            inner = createDataSource( maxPoolSize, connectionTimeoutMs );
+        }
+        catch ( HikariPool.PoolInitializationException e )
+        {
+            throw new IllegalStateException( "Unable to create pool", e );
+        }
         return new JfrDataSource( inner ); // Monitor JDBC traffic with JFR: #61680
     }
 

